### PR TITLE
Outsource data channels

### LINKED ISSUE #1: Integration tests
I'm not yet sure whether it's possible to create instrumentation tests without the main application being an app with activities. We need access to a context though.

An alternative would be to create a separate project as a test project. Could also act as a demo.

### LINKED ISSUE #26: Outsource data channels
Disband tight interaction between the task and WebRTC data channels.

Can be implemented simliar to https://github.com/saltyrtc/saltyrtc-task-webrtc-js/pull/28

### PR DESCRIPTION
**Pull request by lgrahl**
Disband tight interaction between the task and WebRTC data channels

`DataChannel` instances are no longer handled by the task. The application has to take care of managing data channels itself. If the application wants to leverage the established SaltyRTC session keys, it can use a DataChannelCryptoContext instance to encrypt and decrypt data.

Furthermore, if the application wants to hand over the signalling channel to a dedicated data channel, that data channel must also be managed by the application itself while the `SignalingTransport` will take care of the crypto and the chunking. Concretely, the application can request a `ISignalingTransportLink` instance which contains all necessary information to create a data channel for the purpose of handing over the signalling transport. The application then has to provide an implementation of a `ISignalingTransportHandler` towards the task. As a rule of thumb, the `ISignalingTransportLink` handles all incoming messages and events forward by the application while the `ISignalingTransportHandler` handles all outgoing messages and events forwarded by the task.

Resolves #1
Resolves #26

To do:

- [x] Define new API
- [x] Update tests
- [x] Add new tests
- [x] Update readme/docs
  - API documentation
  - Testing
  - Coding guidelines, ...
- [x] Ensure compatibility with legacy `v0`
  - Update the demo for that purpose: saltyrtc/saltyrtc-demo#6

### CODE REVIEW

File: src/main/java/org/saltyrtc/tasks/webrtc/transport/ISignalingTransportHandler.java
Comment (dbrgn) [inline src/main/java/org/saltyrtc/tasks/webrtc/transport/ISignalingTransportHandler.java]: So far in the Java client and task codebases we never use the `I` prefix notation for interfaces. If there isn't a strong reason to do it differently in this case, I would drop the prefix.

File: README.md
Comment (dbrgn) [inline README.md:108]: Would that even be possible? The handover is done by the client, so there are no handover messages that the user could accidentally encrypt and chunk twice, right?

File: README.md
Comment (dbrgn) [inline README.md:235]: Good idea copying over this script :slightly_smiling_face:

File: README.md
Comment (dbrgn) [inline README.md]:
```suggestion
    cd saltyrtc-server-python
```

File: build.gradle
Comment (dbrgn) [inline build.gradle:30]: I guess there's no such thing as a "test repository", right?

File: src/main/java/org/saltyrtc/tasks/webrtc/events/MessageHandler.java
Comment (dbrgn) [inline src/main/java/org/saltyrtc/tasks/webrtc/events/MessageHandler.java:32]: Hmm, for some reason the two lines above use tabs for indenting. Maybe you can fix that.

File: src/test/java/org/saltyrtc/tasks/webrtc/tests/crypto/DataChannelCryptoContextTest.java
Comment (dbrgn) [inline src/test/java/org/saltyrtc/tasks/webrtc/tests/crypto/DataChannelCryptoContextTest.java:72]: Wow, this JUnit5 stuff is quite nice!

**Review by dbrgn — APPROVED** [review]
I didn't read every single line, but looks good overall.

Only change request (besides the typo) would be the interface notation.

**Comment by dbrgn** [comment]
@lgrahl what's the status on this?

**Comment by lgrahl** [comment]
Leave it open for now. I was in the middle of writing the benchmark demo in the SaltyRTC demo project before merging this. Need to continue on that soonish.

File: src/main/java/org/saltyrtc/tasks/webrtc/messages/Candidate.java
Comment (dbrgn) [inline src/main/java/org/saltyrtc/tasks/webrtc/messages/Candidate.java]: Should we call this field (and the getter) `sdp` to match the WebRTC `IceCandidate` constructor?

    public IceCandidate(String sdpMid, int sdpMLineIndex, String sdp)

Would also match `Offer` and `Answer`.

File: src/main/java/org/saltyrtc/tasks/webrtc/messages/Candidate.java
Comment (dbrgn) [inline src/main/java/org/saltyrtc/tasks/webrtc/messages/Candidate.java:30]: This is nullable (as specified by the protocol), but the `sdpMLineIndex` parameter of the `IceCandidate` constructor wants an `int`. Any suggestion what to do if a candidate without an `sdpMLineIndex` would arrive? Throw away the candidate, or can this value be faked somehow? (I don't really know what `sdpMLineIndex` does.)

File: README.md
Comment (dbrgn) [inline README.md:223]: In Threema we're on minSdkVersion 16, btw :wink:

File: README.md
Comment (dbrgn) [inline README.md:174]: Example does not handle `UntiedException` in closing() and closed().

File: README.md
Comment (dbrgn) [inline README.md:184]: Example does not handle `UntiedException`.

File: README.md
Comment (dbrgn) [inline README.md:90]: Should we mention somewhere that the crypto context should only be retrieved once? Should we maybe even return a warning or error if the user calls `getCryptoContext` with the same data channel id twice?

When trying to get this code to run, I initially thought that the state was inside the task and that `getCryptoContext` is just a getter that can be called multiple times. That was wrong of course, but the mistake could happen again to others.

Maybe the method could also be called `createCryptoContext` instead, to make it clear that new state is created.

File: README.md
Comment (lgrahl) [reply README.md:108]: The application could create a crypto context itself before the handover took place. It can also technically use the same channel.

Comment (dbrgn) [reply README.md:108]: Ok. I still think that for most users putting this note in the README will probably be confusing and not too relevant (since the client handles handover), but I leave it to you whether to remove it or not.

File: build.gradle
Comment (lgrahl) [reply build.gradle:30]: I haven't looked into it much but since it's a `testImplementation`, I'd assume it doesn't matter and is not being pulled in?

File: src/main/java/org/saltyrtc/tasks/webrtc/events/MessageHandler.java
Comment (lgrahl) [reply src/main/java/org/saltyrtc/tasks/webrtc/events/MessageHandler.java:32]: That bloody tab/spaces issue gets me every time. :grin:

File: src/main/java/org/saltyrtc/tasks/webrtc/transport/ISignalingTransportHandler.java
Comment (lgrahl) [reply src/main/java/org/saltyrtc/tasks/webrtc/transport/ISignalingTransportHandler.java]: There's `ISignalingTransportLink` and `SignalingTransportLink`. Irgendeinen Tod müssen wir hier sterben. :sweat_smile:

File: src/main/java/org/saltyrtc/tasks/webrtc/messages/Candidate.java
Comment (lgrahl) [reply src/main/java/org/saltyrtc/tasks/webrtc/messages/Candidate.java]: If anything, should probably be renamed to `candidateSdp`. But yeah, for consistency, just `sdp` for now...

File: build.gradle
Comment (dbrgn) [reply build.gradle:30]: The repository would still remain and could potentially override dependencies, I think. But yes, in practice it doesn't really matter.

File: src/main/java/org/saltyrtc/tasks/webrtc/messages/Candidate.java
Comment (lgrahl) [reply src/main/java/org/saltyrtc/tasks/webrtc/messages/Candidate.java:30]: The index refers to the media line. IIRC it's only used if you don't provide a `mid` which is preferred. So, it looks up the corresponding media line by the `mid` and falls back to the index of the media line (within the remote description's SDP) if not provided.

Comment (lgrahl) [reply src/main/java/org/saltyrtc/tasks/webrtc/messages/Candidate.java:30]: For data channels, this is always `0` unless someone tries to negotiate audio/video with the web client...

File: README.md
Comment (lgrahl) [reply README.md:223]: Yup but IIRC we can still use API level 24?

File: build.gradle
Comment (lgrahl) [reply build.gradle:30]: If ordering matters, I can reverse if you want...

File: README.md
Comment (lgrahl) [reply README.md:174]: Mh, true, but do you feel it necessary to add this to the readme? It's not a `RuntimeException`, so I don't mind not catching it in the example.

File: src/main/java/org/saltyrtc/tasks/webrtc/transport/ISignalingTransportHandler.java
Comment (dbrgn) [reply src/main/java/org/saltyrtc/tasks/webrtc/transport/ISignalingTransportHandler.java]: Will there be multiple implementations of the `ISignalingTransportLink` interface, or would a single public class suffice? The API is `task.getTransportLink()` does not require an interface. And I don't think a library user will ever provide a custom implementation, right?

File: README.md
Comment (lgrahl) [reply README.md:90]: > Should we mention somewhere that the crypto context should only be retrieved once?

It's actually fine to do so since the cookie is different every time.

> When trying to get this code to run, I initially thought that the state was inside the task and that `getCryptoContext` is just a getter that can be called multiple times. That was wrong of course, but the mistake could happen again to others.
> 
> Maybe the method could also be called `createCryptoContext` instead, to make it clear that new state is created.

I'll rename it. :+1:

File: README.md
Comment (dbrgn) [reply README.md:223]: If a runtime feature is only supported in API level 24, and not in API level 16, then we cannot use it.

Many of the Java 8 features are not in the runtime though (e.g. lambdas).

File: src/main/java/org/saltyrtc/tasks/webrtc/transport/ISignalingTransportHandler.java
Comment (lgrahl) [reply src/main/java/org/saltyrtc/tasks/webrtc/transport/ISignalingTransportHandler.java]: Will think about that for a second but I think you're right.


... Until we meet the *same name* problem again, then. :smirk_cat:

File: README.md
Comment (dbrgn) [reply README.md:174]: `UntiedException` extends `Exception` and is thus a checked exception which must be caught.

Subtypes of `Error` and `RuntimeException` are unchecked, all other subtypes of `Throwable` are checked.

File: README.md
Comment (lgrahl) [reply README.md:223]: I copy pasted that from https://github.com/saltyrtc/saltyrtc-client-java, so I don't mind changing it but it needs to be done consistently.

File: README.md
Comment (lgrahl) [reply README.md:174]: Yeah, I get it, but that wasn't my point. Does it matter for you to understand the example? Because if not, it will not run without catching it and so it's obvious that it needs to be caught.

File: README.md
Comment (dbrgn) [reply README.md:223]: Yep. I think the comment stems from the fact that API level 24 added some Java 8 runtime features, so it's technically correct. Java 7 OTOH was fully supported for a long time (<16). I think we can keep it this way.

File: README.md
Comment (dbrgn) [reply README.md:174]: I think it would be nice if examples would act as "best practice" and could actually be copy-pasted. Otherwise someone getting started with the codebase will be confused by the `UntiedException` and will not know how to handle it without diving into the codebase.

Comment (lgrahl) [reply README.md:174]: It would just clutter the ~~example~~guide. There are more places where you'd have to catch that. Generally agree with copy/pastable examples... but not in the readme (IMO). This is more of a guide, not a truly working example. The demo is an example. This isn't.

Edit: Enough edits.

File: build.gradle
Comment (lgrahl) [reply build.gradle:30]: Should be fine:

> A project can have multiple repositories. Gradle will look for a dependency in each repository in the order they are specified, stopping at the first repository that contains the requested module.

https://docs.gradle.org/current/userguide/dependency_management_for_java_projects.html#N10621

**Comment by lgrahl** [comment]
The only blocker left is the discussion regarding the guide in the readme.

File: README.md
Comment (dbrgn) [reply README.md:108]: Quick "Verständnisfrage" to make sure... Creating a crypto context simply initializes some keys, but has no side effects on the signaling channel itself, right?

File: README.md
Comment (dbrgn) [reply README.md:174]: Not terribly user-friendly for people new to this library (since there *are* no docs besides the README), but that's not our main concern right now, so fair enough.

File: README.md
Comment (dbrgn) [reply README.md:90]: > It's actually fine to do so since the cookie is different every time.

It's fine security-wise, but decrypting will fail.

But since you renamed the method (`createCryptoContext`) it should be obvious that creating another crypto context will not result in the same data being used, so I'd consider this resolved.

File: README.md
Comment (lgrahl) [reply README.md:108]: Yes. The point of the comment is just that you don't need to do crypto on the handed over signalling channel which may not be 100% clear.

**Pull request merged**
